### PR TITLE
Migrate React-Redux Curriculum into FCC Codebase

### LINKED ISSUE #16115: [beta] Add React and Redux challenges to seed directory once they're working

### PR DESCRIPTION
**Pull request by bonham000**
#### Key:

- :white_check_mark: **Challenge is migrated and solution has been successfully QA'd in FCC editor.**
- :1st_place_medal: **Challenge copy has been QA'd and confirmed, ready to ship.**

### React Challenges:
- :white_check_mark: :1st_place_medal: **1:** Create a Simple JSX element
- :white_check_mark: :1st_place_medal: **2:** Create a Complex JSX Element
- ✅ :1st_place_medal: **3:** Add Comments in JSX
- ✅ :1st_place_medal: **4:** Render HTML Elements to the DOM
- ✅ :1st_place_medal: **5:** Define an HTML Class in JSX
- ✅ :1st_place_medal: **6:** Learn About Self-Closing JSX Tags
- ✅ :1st_place_medal: **7:** Create a Stateless Functional Component
- ✅ :1st_place_medal: **8:** Create a Component with React 
- ✅ :1st_place_medal: **9:** Create a Component with Composition
- ✅ :1st_place_medal: **10:** Use React to Render Nested Components
- ✅ :1st_place_medal: **11:** Compose React Components
- ✅ :1st_place_medal: **12:** Render a Class Component to the DOM
- ✅ :1st_place_medal: **13:** Write a React Component from Scratch
- ✅ :1st_place_medal: **14:** Pass a String to a Functional Component
- ✅ :1st_place_medal: **15:** Pass an Array as Props
- ✅ :1st_place_medal: **16:** Use Default Props
- ✅ :1st_place_medal: **17:** Override Default Props
- ✅ :1st_place_medal: **18:** Use PropTypes to Define the Props You Expect
- ✅ :1st_place_medal: **19:** Access Props Using this.props 
- ✅ :1st_place_medal: **20:** Using Props with Stateless Functional Components
- ✅ :1st_place_medal: **21:** Create a Stateful Component 
- ✅ :1st_place_medal: **22:** Render State in the UI 
- ✅ :1st_place_medal: **23:** Render State in the UI Another Way
- ✅ :1st_place_medal: **24:** Set State with this.setState
- ✅ :1st_place_medal: **25:** Bind 'this' to a Class Method
- ~**26:** Bind 'this' with an ES6 Arrow Function~
- ✅ :1st_place_medal: **27:** Use State to Toggle an Element
- ✅ :1st_place_medal: **28:** Write a Simple Counter
- ✅ :1st_place_medal: **29:** Create a Controlled Input
- ✅ :1st_place_medal: **30:** Create a Controlled Form
- ✅ :1st_place_medal: **31:** Pass State as Props to Child Components
- ✅ :1st_place_medal: **32:** Pass a Callback as Props
- ✅ :1st_place_medal: **33:** Use the Lifecycle Method componentWillMount
- ✅ :1st_place_medal: **34:** Use the Lifecycle Method componentDidMount
- ✅ :1st_place_medal: **35:** Add Event Listeners
- ✅ :1st_place_medal:**36:** Manage Updates with Lifecycle Methods
- ✅ :1st_place_medal: **37:** Optimize Re-Renders with shouldComponentUpdate
- ✅ :1st_place_medal: **38:** Introducing Inline Styles
- ✅ :1st_place_medal: **39:** Add Inline Styles in React
- ✅ :1st_place_medal: **40:** Use Advanced JavaScript in React Render Method
- ✅ :1st_place_medal: **41:** Render with an If/Else Condition
- ✅ :1st_place_medal: **42:** Use && for a More Concise Conditional
- ~**43:** Return null to Prevent Rendering~
- ✅ :1st_place_medal: **44:** Use a Ternary Expression for Conditional Rendering
- ✅ :1st_place_medal: **45:** Render Conditionally from Props __Note: Pete, refactor with `simulate()`__
- ✅ :1st_place_medal: **46:** Change Inline CSS Conditionally Based on Component State
- ✅ :1st_place_medal: **47:** Use Array.map() to Dynamically Render Elements
- ✅ :1st_place_medal: **48:** Give Sibling Elements a Unique Key Attribute
- ✅ :1st_place_medal: **49:** Use Array.filter() to Dynamically Filter an Array
- ✅ :1st_place_medal: **50:** Render React on the Server with renderToString

### Redux Challenges:

- ✅ :1st_place_medal: **1:** Create a Redux Store
- ✅ :1st_place_medal: **2:** Get the State of the Redux Store
- ✅ :1st_place_medal: **3:** Define a Redux Action
- ✅ :1st_place_medal: **4:** Define an Action Creator
- ✅ :1st_place_medal: **5:** Dispatch an Action Event
- ✅ :1st_place_medal: **6:** Handle an Action in the Store
- ✅ :1st_place_medal: **7:** Use a Switch Statement to Handle Multiple Actions
- ✅ :1st_place_medal: **8:** Use const for Action Types
- ✅ :1st_place_medal: **9:** Register a Store Listener
- ✅ :1st_place_medal: **10:** Combine Multiple Reducers
- ✅ :1st_place_medal: **11:** Send Action Data to the Store
- ✅ :1st_place_medal: **12:** Use Middleware to Handle Asynchronous Actions
- ✅ :1st_place_medal: **13:** Write a Counter with Redux
- ✅ :1st_place_medal: **14:** Never Mutate State
- ✅ :1st_place_medal: **15:** Use the Spread Operator on Arrays
- ✅ :1st_place_medal: **16:** Remove an Item from an Array
- ✅ :1st_place_medal: **17:** Copy an Object with Object.assign
- ~**18:** Use the ES6 Spread Operator with Objects~

### React-Redux Challenges:

- ✅ :1st_place_medal: **1:** Getting Started with React Redux
- ✅ :1st_place_medal: **2:** Manage State Locally First
- ✅ :1st_place_medal: **3:** Extract State Logic to Redux
- ✅ :1st_place_medal: **4:** Use Provider to Connect Redux to React
- ✅ :1st_place_medal: **5:** Map State to Props
- ✅ :1st_place_medal: **6:** Map Dispatch to Props
- ✅ :1st_place_medal: **7:** Connect Redux to React
- ✅ :1st_place_medal: **8:** Connect Redux to the Messages App
- ✅ :1st_place_medal: **9:** Extract Local State into Redux
- ✅ :1st_place_medal: **10:** Moving Forward From Here

***

### Notes
* There is no support for `async` tests in the automated test environment (we skip them currently).
* These challenges rely on `editor.getOriginalCode()` to provide the original, untranspiled code string.
* Transpiled `head` and `tail` code is required in the FCC editor during for some of these challenges but is currently not working.
* Preview function is currently not working for these challenges.

***

### Type of Change
- [x] New feature (non-breaking change which adds new functionality)
- [x] Closes #16115

### Description
Migrate the React-Redux Curriculum from the [prototype module](https://github.com/bonham000/fcc-react-tests-module) into the FCC codebase.

### Checklist
- [x] Populate challenge files (title, description, seed code, solution).
- [x] Challenge solutions pass tests under automated test suite.
- [x] Manually check solutions work in app.
- [x] QA challenge & test copy

***

### Motivational Images

screen shot 2017-12-03 at 8 10 11 pm

screen shot 2017-12-03 at 7 59 26 pm

screen shot 2017-12-11 at 8 02 36 pm

### CODE REVIEW

**Comment by Ethan-Arrowood** [comment]
@bonham000 I'm interested in helping you move over the react and redux challenges. Let me know how I can help!

**Comment by bonham000** [comment]
Sounds good @Ethan-Arrowood I sent you a message on Gitter.

**Comment by anhtran1906** [comment]
@bonham000 Hi I'm also interested in helping you migrate the react/redux challenges over also. Please let me know how I could help

**Comment by no-stack-dub-sack** [comment]
Hi @anhtran1906, thanks for reaching out! What time zone are you in and what is your schedule like? If you are on Gitter (which will make communication much faster), send one or both of us a message soon, and we can add you to a room where we can discuss how you can help. Thanks again!

**Comment by anhtran1906** [comment]
@no-stack-dub-sack Hi I'm in EST timezone and my schedule for the week from December 16th is pretty empty. I am on Gitter as @anhtran1906 also. Please add me in a room so we could discuss more! Thank you!

**Comment by Ethan-Arrowood** [comment]
@anhtran1906 welcome to the crew I just added you to the gitter chat

**Comment by Ethan-Arrowood** [comment]
Going to use this comment to track my textual QA of the migrated challenges. 

Completed Challenges: [ ]

**Comment by bonham000** [comment]
@Ethan-Arrowood sounds great! Thanks for QA'ing them. Keep updating here and we'll add medals :1st_place_medal: to the passing challenges in the comment above.

We've finished migrated all the challenges so this final review is that last checklist item. Thanks again for your help.

**Comment by Ethan-Arrowood** [comment]
@bonham000 I'll get through all of them today (in batches). 🎉

**Comment by Ethan-Arrowood** [comment]
All React challenges are QA'd Redux and React-Redux coming this evening.

**Comment by Ethan-Arrowood** [comment]
Redux done. React-Redux up next 💯

File: client/frame-runner.js [hunk 1]
```diff
@@ -29,7 +52,10 @@ document.addEventListener('DOMContentLoaded', function() {
 
   document.__runTests = function runTests(tests = []) {
     /* eslint-disable no-unused-vars */
-    const editor = { getValue() { return source; } };
+    const editor = {
+      getValue() { return source; },
+      getOriginalCode() { return originalCode; }
+    };
```
Comment (BerkeleyTrue) [inline client/frame-runner.js]: The tests that require access to user code should use `getUserInput` see: https://github.com/freecodecamp/freecodecamp/blob/staging/seed/challenges/03-front-end-libraries/react.json#L65

File: common/app/routes/Challenges/redux/execute-challenge-epic.js [hunk 2]
```diff
@@ -117,7 +116,15 @@ export function executeChallengeEpic(actions, { getState }, { document }) {
...
           return buildFromFiles(files, required, false)
-            .do(frameTests)
+            .do(build => {
+              // Hack in original code for tests:
+              let originalCode;
+              if (files && files.indexjsx && files.indexjsx.contents) {
+                originalCode = files.indexjsx.contents;
+                build.sources.originalCode = originalCode;
+              }
+              return frameTests(build);
```
Comment (BerkeleyTrue) [inline common/app/routes/Challenges/redux/execute-challenge-epic.js]: see above comment in frame.js

File: common/app/routes/Challenges/redux/index.js [hunk 3]
```diff
@@ -113,7 +113,7 @@ export const classicEditorUpdated = createAction(
 export const modernEditorUpdated = createAction(
   types.modernEditorUpdated,
   null,
-  createFilesMetaCreator
+  updateFileMetaCreator
```
Comment (BerkeleyTrue) [inline common/app/routes/Challenges/redux/index.js]: This change has already landed in staging

File: common/app/routes/Challenges/utils/frame.js [hunk 4]
```diff
@@ -101,6 +101,7 @@ const writeTestDepsToDocument = frameReady => ctx => {
   // default for classic challenges
   // should not be used for modern
   tests.__source = sources['index'] || '';
+  tests.__originalCode = sources['originalCode'] || '';
```
Comment (BerkeleyTrue) [inline common/app/routes/Challenges/utils/frame.js]: see above comment in framejs

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by BerkeleyTrue** [comment]
This is continued in PR #16200

**Pull request closed without merging**